bindings/rust/src/lazy_reader.rs

```diff
@@ -0,0 +1,518 @@
+            reader,
+            total_size,
+            cache_start_point: 0,
+            cache_size: 0, // when created, cache is not filled
+            cache: vec![0u8; MAX_CACHE_SIZE],
+        };
+        Cursor {
+            offset: 0,
+            size: total_size,
+            data_source: Rc::new(RefCell::new(data_source)),
+        }
+    }
+    pub fn read_at(&self, buf: &mut [u8]) -> Result<usize, Error> {
+        let read_len = min(self.size, buf.len());
+        let ds = &mut *self.data_source.borrow_mut();
+        if read_len > ds.cache.len() {
+            return ds.reader.read(buf, self.offset);
+        }
+        if self.offset < ds.cache_start_point
+            || (self.offset + read_len) > (ds.cache_start_point + ds.cache_size)
+        {
+            let reader = &ds.reader;
+            let size = reader.read(&mut ds.cache[..], self.offset)?;
+            if size < read_len {
+                return Err(Error::Read(format!(
+                    "read_at: `if size({}) < read_len({})`",
+                    size, read_len
+                )));
+            }
+            ds.cache_size = size;
+            ds.cache_start_point = self.offset;
+
+            if ds.cache_size > ds.cache.len() {
+                return Err(Error::Read(format!(
+                    "read_at: `if ds.cache_size({}) > ds.cache.len()({})`",
+                    ds.cache_size,
+                    ds.cache.len()
+                )));
+            }
+        }
+        if self.offset < ds.cache_start_point
+            || (self.offset - ds.cache_start_point) > ds.cache.len()
+        {
+            return Err(Error::Read(format!(
+                "read_at: `if self.offset({}) < ds.cache_start_point({}) || ...`",
+                self.offset, ds.cache_start_point
+            )));
+        }
+        let read_point = self.offset - ds.cache_start_point;
+        if read_point + read_len > ds.cache_size {
+            return Err(Error::Read(format!(
+                "read_at: `if read_point({}) + read_len({}) > ds.cache_size({})`",
+                read_point, read_len, ds.cache_size
+            )));
+        }
+        buf.copy_from_slice(&ds.cache[read_point..(read_point + read_len)]);
+        Ok(read_len)
+    }
+
+    pub fn add_offset(&mut self, offset: usize) -> Result<(), Error> {
+        self.offset = self.offset.checked_add(offset).ok_or_else(|| {
+            Error::Overflow(format!(
+                "add_offset: self.offset({}) + offset({})",
+                self.offset, offset
+            ))
+        })?;
+        Ok(())
+    }
+
+    pub fn sub_size(&mut self, shrink_size: usize) -> Result<(), Error> {
+        self.size = self.size.checked_sub(shrink_size).ok_or_else(|| {
+            Error::Overflow(format!(
+                "sub_size: self.size({}) - shrink_size({})",
+                self.size, shrink_size
+            ))
+        })?;
+        Ok(())
+    }
+
+    pub fn validate(&self) -> Result<(), Error> {
```

Comment:
It might be a good idea to revisit here, and distinguish functions that are used internally, and externally used functions. It really seems to me that this validate function, is used internally in the Cursor? In that sense there is not need to put that `pub` here.

And there might also be other methods in this structure that are only required internally, so we might want to think about those.

Reply:
Sometime we need to adjust cursor manually, see c code: https://github.com/XuJiandong/omnilock/blob/90ad8ba6744b9c5d261c713b6b71c30740d54a98/c/cobuild.c#L269-L271

Reply:
Like the comment below, I don't really see any operation that cannot be covered by a slicing operation. We can provide a single slicing operation for cutting a cursor into any desirable range. Then the only place we need to do validation, is within the slicing operation.